### ISSUE TITLE
Store a configuration file for each device

### ISSUE THREAD

**Issue by felixlapierre** [bean-pod/switchboard#261]
After a discussion with the PM the requirements have shifted a bit:

- Store a file containing the configuration of each device (not each decoder). The format of the file is not assumed to be JSON, it could be any kind of file, and we will not perform any validation on it.
- We will not store a configurationSchema since we will allow configurations to have any file format. We will leave it up to the device to reject a configuration if it does not like it.
- Allow devices to register without providing a starting configuration.
- Allow users to call an endpoint to upload a new configuration file which will replace the old one.
- Allow users to call an endpoint to retrieve the current configuration
- Track the time at which the configuration was last modified. Make sure this value cannot be "manually" changed, it should be set to the current time whenever the config is changed.
- Store the config as a binary file in the db to disallow database injection.

**Comment by mohhef** [comment]
@felixlapierre 
i have some questions
1. "properties exist for the configuration of that device". whats a configuration?, also in other instances im not sure whats a configuration.
2. im a little confused on whats this task and what are we trying to acheive?

**Comment by felixlapierre** [comment]
Here's a description of the use case:

Suppose we have a video sender. It supports three different resolutions: 1080p, 720p and 360p. The encoder could be manually configured onsite by an operator to use the right resolution, but we want to make it easier by letting the user select the resolution in the cloud.

However, every device is different: some support 4K also, some might only support 1080 or 720, some might not actually have a configurable resolution. Some might have other stuff that can be configured (like for example the bitrate). Rather than code a bunch of properties in our device database entity, we need a solution that works for all kinds of configuration.

I envisioned a solution where each device tells us what properties it has that are configurable, so we can be flexible to the needs of any device. I figured we can do this by storing 2 things for each device

1. A JSON schema explaining what properties the device's configuration has, what types they are, what the default values are.
2. A JSON object with the current configuration.

I envision that the device will send us its JSON schema, and the JSON object with its current configuration, when it registers.

In the frontend, we will acquire this schema and use it to generate a dynamic component with an editable field for each property, that validates the inputted data based on the info in the schema. If the user changes values, it will send a PUT request to the server to update the config.

When devices "poll" the server (to get the list of streams they need to do) they will also get the datetime at which their config was last updated. When they see that datetime change, they will be responsible for pulling the new config from the database.

Let me know if this answers your questions!
